olp-cpp-sdk-dataservice-read/src/repositories/PartitionsRepository.cpp

```diff
@@ -458,6 +463,119 @@ CancellationToken PartitionsRepository::GetPartitionsById(
+  partition_request.WithLayerId(layer)
+      .WithBillingTag(data_request.GetBillingTag())
+      .WithVersion(version);
+
+  auto fetch_option = data_request.GetFetchOption();
+
+  if (fetch_option != OnlineOnly) {
+    auto cached_partitions = repository.Get(partition_request, partitions);
+    if (cached_partitions.GetPartitions().size() == partitions.size()) {
+      auto key = data_request.CreateKey();
+      OLP_SDK_LOG_INFO_F(kLogTag, "cache data '%s' found!", key.c_str());
+      return cached_partitions;
+    } else if (fetch_option == CacheOnly) {
+      auto key = data_request.CreateKey();
+      OLP_SDK_LOG_INFO_F(kLogTag, "cache catalog '%s' not found!", key.c_str());
+      return ApiError(ErrorCode::NotFound,
+                      "Cache only resource not found in cache (partition).");
+    }
+  }
+
+  auto query_api =
+      ApiClientLookup::LookupApi(catalog, cancellation_context, "query", "v1",
+                                 fetch_option, std::move(settings));
+
+  if (!query_api.IsSuccessful()) {
+    return query_api.GetError();
+  }
+
+  const client::OlpClient& client = query_api.GetResult();
+
+  Condition condition(timeout);
+
+  // when the network operation took too much time we cancel it and exit
+  // execution, to make sure that network callback will not access dangling
+  // references we protect them with atomic bool flag.
+  auto flag = std::make_shared<std::atomic_bool>(true);
+
+  QueryApi::PartitionsResponse query_response;
+  auto query_client_callback = [&,
+                                flag](QueryApi::PartitionsResponse response) {
+    if (flag->exchange(false)) {
+      query_response = std::move(response);
+      condition.Notify();
+    }
+  };
+
+  cancellation_context.ExecuteOrCancelled(
+      [&, flag]() {
+        auto token = QueryApi::GetPartitionsbyId(
+            client, layer, partitions, version, boost::none,
+            data_request.GetBillingTag(), std::move(query_client_callback));
+        return client::CancellationToken([&, token, flag]() {
+          if (flag->exchange(false)) {
+            token.cancel();
+            condition.Notify();
+          }
+        });
+      },
+      [&]() { condition.Notify(); });
+
+  if (!condition.Wait()) {
+    cancellation_context.CancelOperation();
+    return client::ApiError(client::ErrorCode::RequestTimeout,
+                            "Network request timed out.");
+  }
+
+  flag->store(false);
+
+  if (cancellation_context.IsCancelled()) {
+    // We can't use api response here because it could potentially be
+    // uninitialized.
+    return client::ApiError(client::ErrorCode::Cancelled,
+                            "Operation cancelled.");
+  }
+
+  if (query_response.IsSuccessful()) {
+    auto key = data_request.CreateKey();
+    OLP_SDK_LOG_INFO_F(kLogTag, "put '%s' to cache", key.c_str());
+    repository.Put(partition_request, query_response.GetResult(),
+                   0 /* TODO: expiration */);
```

Comment:
```suggestion
                   boost::none /* TODO: expiration */);
```
Maybe better to use `boost::none` in this case?